signature/cose/envelope.go

```diff
@@ -0,0 +1,595 @@
+	}
+	signerInfo.SignedAttributes.SigningTime = time.Unix(signingTime, 0)
+
+	// populate signerInfo.SignedAttributes.Expiry
+	if exp, ok := protected[headerLabelExpiry]; ok {
+		expiry, ok := exp.(int64)
+		if !ok {
+			return &signature.InvalidSignatureError{Msg: "expiry requires int64 type"}
+		}
+		signerInfo.SignedAttributes.Expiry = time.Unix(expiry, 0)
+	}
+
+	// populate signerInfo.SignedAttributes.ExtendedAttributes
+	signerInfo.SignedAttributes.ExtendedAttributes, err = generateExtendedAttributes(extendedAttributeKeys, protected)
+	return err
+}
+
+// validateCritHeaders does a two-way check, namely:
+// 1. validate that all critical headers are present in the protected bucket
+// 2. validate that all required headers(as per spec) are marked critical
+func validateCritHeaders(protected cose.ProtectedHeader) ([]string, error) {
+	// This ensures all critical headers are present in the protected bucket.
+	labels, err := protected.Critical()
+	if err != nil {
+		return nil, err
+	}
+
+	// set of headers that must be marked as crit
+	mustMarkedCrit := make(map[interface{}]struct{})
+	mustMarkedCrit[headerLabelSigningScheme] = struct{}{}
+	signingScheme, ok := protected[headerLabelSigningScheme].(string)
+	if !ok {
+		return nil, &signature.InvalidSignatureError{Msg: "malformed signingScheme"}
+	}
+	if signature.SigningScheme(signingScheme) == signature.SigningSchemeX509SigningAuthority {
+		mustMarkedCrit[headerLabelAuthenticSigningTime] = struct{}{}
+	}
+	if _, ok := protected[headerLabelExpiry]; ok {
+		mustMarkedCrit[headerLabelExpiry] = struct{}{}
+	}
+
+	// validate that all required headers(as per spec) are marked as critical
+	for _, label := range labels {
+		delete(mustMarkedCrit, label)
+	}
+	if len(mustMarkedCrit) != 0 {
+		headers := make([]interface{}, 0, len(mustMarkedCrit))
+		for k := range mustMarkedCrit {
+			headers = append(headers, k)
+		}
+		return nil, &signature.InvalidSignatureError{Msg: fmt.Sprintf("these required headers are not marked as critical: %v", headers)}
+	}
+
+	// fetch all the extended signed attributes
+	systemHeaders := []interface{}{cose.HeaderLabelAlgorithm, cose.HeaderLabelCritical, cose.HeaderLabelContentType,
+		headerLabelExpiry, headerLabelSigningScheme, headerLabelSigningTime, headerLabelAuthenticSigningTime}
+	var extendedAttributeKeys []string
+	for label := range protected {
+		if contains(systemHeaders, label) {
+			continue
+		}
+		label, ok := label.(string)
+		if !ok {
+			return nil, &signature.InvalidSignatureError{Msg: "extendedAttributes key requires string type"}
+		}
+		extendedAttributeKeys = append(extendedAttributeKeys, label)
+	}
+
+	return extendedAttributeKeys, nil
+}
+
+// generateExtendedAttributes generates []signature.Attribute during SignerInfo process
+func generateExtendedAttributes(extendedAttributeKeys []string, protected cose.ProtectedHeader) ([]signature.Attribute, error) {
+	criticalHeaders, ok := protected[cose.HeaderLabelCritical].([]interface{})
```

Comment:
Same. Do we need to check key exists first?

Reply:
> Same. Do we need to check key exists first?

Same here. this line should check both existence and type assertion?